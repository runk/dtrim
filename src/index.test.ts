import _ from "lodash";
import { trimmer as trimmerFactory } from ".";

import fixtureLarge from "./__mocks__/fixture-large";
import fixtureRealWorld from "./__mocks__/fixture-real-world";

describe("trimmer", () => {
  const trimmer = trimmerFactory({
    depth: 3,
    string: 32,
    size: 16
  });

  test("large data sets", () =>
    expect(trimmer(fixtureLarge)).toMatchSnapshot());

  test("real world example", () =>
    expect(trimmer(fixtureRealWorld)).toMatchSnapshot());

  test("circular structures", () => {
    const refA = { foo: "bar" };
    const refB = { refA, something: "else" };
    // @ts-ignore
    refA.backref = refB;
    expect(trimmer(refA)).toMatchSnapshot();
  });

  test("immutability", () => {
    const input = { list: _.range(0, 1024) };
    const trimmed = trimmer(input);
    expect(input.list).toHaveLength(1024);
    expect(trimmed.list).toHaveLength(16 + 1);
  });

  describe("rules", () => {
    test("#string", () => {
      const input = { short: "hi", long: _.repeat("a", 1024) };
      const trimmer = trimmerFactory({ string: 4 });
      expect(trimmer(input)).toEqual({
        short: "hi",
        long: "aaaa..."
      });
    });

    test("#buffer", () => {
      const input = { buf: Buffer.alloc(8) };
      expect(trimmerFactory({ buffer: true })(input)).toEqual({
        buf: "Buffer(8)"
      });

      expect(trimmerFactory({ buffer: false })(input)).toEqual({
        buf: input.buf
      });
    });

    test("#depth", () => {
      const input = {
        deep: _.set({}, "a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.r", "very deep"),
        shallow: _.set({}, "a.b", "quite shallow")
      };
      expect(trimmerFactory({ depth: 3 })(input)).toEqual({
        deep: {
          a: { b: { c: "... deeper levels trimmed" } }
        },
        shallow: {
          a: { b: "quite shallow" }
        }
      });
    });

    test("#size", () => {
      const bigList = _.range(0, 16);
      const smallList = _.range(0, 2);
      const input = {
        bigList,
        bigObject: _.zipObject(bigList, bigList),
        smallList,
        smallObject: _.zipObject(smallList, smallList)
      };

      const output = trimmerFactory({ size: 5 })(input);
      expect(output).toEqual({
        bigList: [0, 1, 2, 3, 4, "... 11 more items"],
        bigObject: { data: "Object(16)" },
        smallList: [0, 1],
        smallObject: { 0: 0, 1: 1 }
      });
    });
  });
});
