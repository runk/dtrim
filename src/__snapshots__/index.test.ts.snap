// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trimmer circular structures 1`] = `
Object {
  "backref": Object {
    "refA": Object {
      "backref": "[Object]",
      "foo": "bar",
    },
    "something": "else",
  },
  "foo": "bar",
}
`;

exports[`trimmer large data sets 1`] = `
Object {
  "bao": NaN,
  "bar": undefined,
  "baz": false,
  "buffer": "Buffer(1024)",
  "deep": Object {
    "a": Object {
      "b": "[Object]",
    },
  },
  "fn": "[Function]",
  "foo": null,
  "list": "Array(1024)",
  "object": "Object(1024)",
  "string": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa...",
}
`;

exports[`trimmer real world example 1`] = `
Object {
  "config": Object {
    "baseURL": "http://example.xyz",
    "data": Object {
      "_boundary": "--------------------------012345...",
      "_currentStream": null,
      "_events": "[Object]",
      "_eventsCount": 0,
      "_overheadLength": 283,
      "_released": true,
      "_streams": "[Object]",
      "_valueLength": 9523,
      "_valuesToMeasure": "[Object]",
      "dataSize": 0,
      "maxDataSize": 2097152,
      "pauseStreams": true,
      "readable": true,
      "writable": false,
    },
    "headers": Object {
      "Accept": "application/json, text/plain, */...",
      "Content-Type": "multipart/form111-data; boundary...",
      "authorization": "Bearer abcdef",
      "user-agent": "curl",
      "x-request-id": "afe950b4-30f0-4dc0-a4c7-349e5c12...",
    },
    "maxContentLength": -1,
    "method": "post",
    "timeout": 5000,
    "transformRequest": Object {},
    "transformResponse": Object {},
    "url": "http://example.xyz/foo/bar",
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
  },
  "request": "Object(34)",
  "response": Object {
    "config": "~config",
    "data": "",
    "headers": "~request~res~headers",
    "request": "~request",
    "status": 500,
    "statusText": "Internal Server Error",
  },
}
`;
